Fix stake split rent-exempt adjustment

CriesofCarrots opened the pull request:
#### Problem
Stake split doesn't properly handle the stake deduction from a Delegated original account when the new account has a balance less than the rent-exempt reserve. The original stake needs to be reduced by the entire balance transferred, except in the case of splitting the entire source stake, when it should simply be zeroed.

Stake split also errors on an attempt to split all of a source account to a new account that holds a balance. This is inconsistent with the behavior for a partial split, where an existing balance is acceptable and used to cover the rent-exempt reserve (see`test_split_with_rent()`).

#### Summary of Changes
- Fix math

Before/after examples added in comments.

This is WIP as I need to coordinate gating with @t-nelson 
cc: @ryoqun @mvines

CriesofCarrots commented:
Splitting partial amount into empty account (default cli behavior)

Before:

```
$ solana stake-account stake.json
Balance: 0.1 SOL
Rent Exempt Reserve: 0.00228288 SOL
Delegated Stake: 0.09771712 SOL
Active Stake: 0.09771712 SOL
Delegated Vote Account Address: 2FPim6EJBTW3uwr36fAhK8BEGB22BjJuxkPbtMYAeVUv

$ solana split-stake stake.json split.json 0.05

Signature: 4u1xhY2NqvsoGQHqh65QvJpT3AVyBoyv2K7Bx6h3JbuJ8NLwMFAeHKCZdhSi8wNm5huYMvmphqHn1nYzs39SLYRw

$ solana stake-account stake.json
Balance: 0.05 SOL
Rent Exempt Reserve: 0.00228288 SOL
Delegated Stake: 0.05 SOL <--- Notice stake is not less the rent-exempt-reserve
Active Stake: 0.05 SOL
Delegated Vote Account Address: 2FPim6EJBTW3uwr36fAhK8BEGB22BjJuxkPbtMYAeVUv

$ solana stake-account split.json`
Balance: 0.05 SOL
Rent Exempt Reserve: 0.00228288 SOL
Delegated Stake: 0.04771712 SOL
Active Stake: 0.04771712 SOL
Delegated Vote Account Address: 2FPim6EJBTW3uwr36fAhK8BEGB22BjJuxkPbtMYAeVUv
```

After:

```
$ solana stake-account stake.json
Balance: 0.1 SOL
Rent Exempt Reserve: 0.00228288 SOL
Delegated Stake: 0.09771712 SOL
Active Stake: 0.09771712 SOL
Delegated Vote Account Address: GGohSWd9vwLDqfu32nXuzYM7B6W8N7baRSo5ph9CJsG9

$ solana split-stake stake.json split.json 0.05

Signature: 4oEubJWX4WcjDQaN2KtN2Li4TsZD3WRDeWfGGAg26FzWATKEbJt8Nc84zhfMDx78pY1oxgnGhQMsWDhB75MJTTSK

$ solana stake-account stake.json
Balance: 0.05 SOL
Rent Exempt Reserve: 0.00228288 SOL
Delegated Stake: 0.04771712 SOL
Active Stake: 0.04771712 SOL
Delegated Vote Account Address: GGohSWd9vwLDqfu32nXuzYM7B6W8N7baRSo5ph9CJsG9

$ solana stake-account split.json
Balance: 0.05 SOL
Rent Exempt Reserve: 0.00228288 SOL
Delegated Stake: 0.04771712 SOL
Active Stake: 0.04771712 SOL
Delegated Vote Account Address: GGohSWd9vwLDqfu32nXuzYM7B6W8N7baRSo5ph9CJsG9
```

CriesofCarrots commented:
Splitting complete stake account into account prefunded with rent-exempt reserve of 0.00228288 SOL (I had to edit cli to use this create-account instruction:
```
system_instruction::create_account(
            authorized_pubkey, // Sending 0, so any signer will suffice
            split_stake_pubkey,
            2282880,
            std::mem::size_of::<StakeState>() as u64,
            &id(),
        )
```

Before:

```
$ solana stake-account stake.json
Balance: 0.1 SOL
Rent Exempt Reserve: 0.00228288 SOL
Delegated Stake: 0.09771712 SOL
Active Stake: 0.09771712 SOL
Delegated Vote Account Address: JCcavabWfWZc9ykWb68Vjvpy6oH6nYWGrkwFdEnk4Z95

$ solana split-stake stake.json split.json 0.1
Error: RPC response error -32002: Transaction simulation failed: Error processing Instruction 1: custom program error: 0x4
```

After:

```
$ solana stake-account /Users/tyeraeulberg/.config/solana/stake.json`
Balance: 0.1 SOL
Rent Exempt Reserve: 0.00228288 SOL
Delegated Stake: 0.09771712 SOL
Active Stake: 0.09771712 SOL
Delegated Vote Account Address: FmM9tLoFrqAAS6K9MrMxDWT2o1RzpE5cv388iYJSMAm2
Stake Authority: Bbqg1M4YVVfbhEzwA9SpC9FhsaG83YMTYoR4a8oTDLX
Withdraw Authority: Bbqg1M4YVVfbhEzwA9SpC9FhsaG83YMTYoR4a8oTDLX

$ solana split-stake stake.json split.json 0.1

Signature: rVtvcTHJX6VYdYBPKyiX9rssVAB9SUSK6jNU28HSmwTMf9HJXVsrKRtrSCfVjbeam8V34CNFR8BmkDwkNPRHFVq

$ solana stake-account stake.json
Error: AccountNotFound: pubkey=CYRJWqiSjLitBAcRxPvWpgX3s5TvmN2SuRY3eEYypFvT

$ solana stake-account split.json
Balance: 0.10228288 SOL
Rent Exempt Reserve: 0.00228288 SOL
Delegated Stake: 0.09771712 SOL
Active Stake: 0.09771712 SOL
Delegated Vote Account Address: FmM9tLoFrqAAS6K9MrMxDWT2o1RzpE5cv388iYJSMAm2
```

CriesofCarrots commented:
@rwalker-com , would you be able to take a look at this and see if I'm missing anything? 🙏

rwalker-com reviewed (approved):
lgtm.  all I had to do was read tests:-)

t-nelson commented on programs/stake/src/stake_state.rs:
```diff
@@ -766,13 +770,31 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+                    let (remaining_stake_delta, split_stake_amount) = if lamports
+                        == self.lamports()?
+                    {
+                        // If split amount equals the full sounce stake, the new split stake must
```
```suggestion
                        // If split amount equals the full source stake, the new split stake must
```

t-nelson commented on programs/stake/src/stake_state.rs:
```diff
@@ -766,13 +770,31 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+                    {
+                        // If split amount equals the full sounce stake, the new split stake must
+                        // equal the same amount, regardless of any current lamport balance in the
+                        // split account. Since split accounts retain the state of their sournce
```
```suggestion
                        // split account. Since split accounts retain the state of their source
```

CriesofCarrots replied:
🤦‍♀️ How many wrong ways can I spell that word?

t-nelson replied:
I believe company policy is to blame the Macbook Pro keyboard!

t-nelson reviewed (approved):
LGTM!  Just a couple comment typoes

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -766,13 +770,31 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+                    let split_stake = stake.split(remaining_stake_delta, split_stake_amount)?;
 
                     self.set_state(&StakeState::Stake(meta, stake))?;
                     split.set_state(&StakeState::Stake(meta, split_stake))?;
```
@CriesofCarrots oops, I found another issue. Namely, the split account's rent-reserve can be set incorrectly. I'm still reading the code, but we could never safely `set_state` identical  `meta` to two stake accounts. We have to individually maintain rent-exempt reserves. another head-ache...

Very unfortunately, the are various sizes of stake accounts on the wild:

```
$ ./target/release/solana stake-account C75NM6yJMa4uAEfdUYnSmMHfVAUiPVd67rqTKaZdd5Z
Balance: 500000 SOL
Rent Exempt Reserve: 0.00228288 SOL
Delegated Stake: 499999.99771712 SOL
Active Stake: 499999.99771712 SOL
Delegated Vote Account Address: CCxSNvJogH6LWyoiEbG7JfcWybw2FqqCExs5GuemChGr
Stake Authority: so1b2w9fpMdqgzHR2UvW4iqkkc8nig6xTeBjU5HMbjG
Withdraw Authority: FdGYQdiRky8NZzN9wZtczTBcWLYYRXrJ3LMDhqDPn5rM


$ ./target/release/solana stake-account 13CJic1hdT9woLL7ctrKi9rBwBFoqjcF2BNnDbZceDnS
Balance: 25 SOL
Rent Exempt Reserve: 0.02878656 SOL
Delegated Stake: 24.97121344 SOL
Active Stake: 24.97121344 SOL
Delegated Vote Account Address: ateamaZDqNWDztxnVKZhRsp4ac53KvT1rVKyU5LnL6o
Stake Authority: 55c8bMTU311mv8KkAnTMtdGGKWAueAGhNUYdVfvjuvNy
Withdraw Authority: 55c8bMTU311mv8KkAnTMtdGGKWAueAGhNUYdVfvjuvNy
```

ryoqun replied:
```
$ ./target/release/solana account C75NM6yJMa4uAEfdUYnSmMHfVAUiPVd67rqTKaZdd5Z | head

Public Key: C75NM6yJMa4uAEfdUYnSmMHfVAUiPVd67rqTKaZdd5Z
Balance: 500000 SOL
Owner: Stake11111111111111111111111111111111111111
Executable: false
Rent Epoch: 107
Length: 200 (0xc8) bytes
0000:   02 00 00 00  80 d5 22 00  00 00 00 00  0d 03 10 73   ......"........s
0010:   44 03 0b f0  4d b2 8a 86  2c 2d 19 93  1f fb c5 e1   D...M...,-......
0020:   50 36 55 ed  9f 03 88 e8  d0 d7 98 bb  d9 4e 19 53   P6U..........N.S

$ ./target/release/solana account 13CJic1hdT9woLL7ctrKi9rBwBFoqjcF2BNnDbZceDnS | head

Public Key: 13CJic1hdT9woLL7ctrKi9rBwBFoqjcF2BNnDbZceDnS
Balance: 25 SOL
Owner: Stake11111111111111111111111111111111111111
Executable: false
Rent Epoch: 107
Length: 4008 (0xfa8) bytes
0000:   02 00 00 00  80 3f b7 01  00 00 00 00  3c 9c cb 27   .....?......<..'
0010:   f9 18 85 f0  46 c2 c1 29  91 56 ab 80  e1 37 8f 30   ....F..).V...7.0
0020:   76 d8 fd ff  ab d7 d0 4b  38 d7 fd ce  3c 9c cb 27   v......K8...<..'
```

ryoqun replied:
That means my upcoming stake account normalization must also adjust rent-exempt reserves.

ryoqun replied:
ref: https://discord.com/channels/428295358100013066/439194979856809985/771378185890693131

t-nelson commented on programs/stake/src/stake_state.rs:
```diff
@@ -761,11 +761,22 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
             match self.state()? {
                 StakeState::Stake(meta, mut stake) => {
                     meta.authorized.check(signers, StakeAuthorize::Staker)?;
+                    let original_lamports_per_byte_year = calculate_lamports_per_byte_year(
```
Do you anticipate this value being used otherwise?  If not, it might simplify things a bit to combine the two utility functions into one `fn calculate_split_rent_exempt_reserve(source: &KeyedAccount, split: &KeyedAccount) -> u64`

CriesofCarrots replied:
No, I hope it won't be used otherwise. Fair point, will squash.

CriesofCarrots replied:
Updated

t-nelson commented on programs/stake/src/stake_state.rs:
```diff
@@ -761,11 +761,22 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+                        ) > meta.rent_exempt_reserve
                     {
                         return Err(InstructionError::InsufficientFunds);
                     }
```
We need to use the split's `rent_exempt_reserve` below, no? (Sorry can't comment outside diff context)

CriesofCarrots replied:
No, the rent_exempt_reserve needs to remain the same between the source and split accounts. We just don't want to allow that to put the split account in a non-exempt state

CriesofCarrots replied:
Mm, maybe we can actually, as long as the stake being split never increases as a result.

CriesofCarrots replied:
Updated. No tests failed, so that's good. I still need to add a case for larger source to smaller split, splitting partial funds.

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -969,6 +1007,17 @@ pub fn calculate_points(
     }
 }
 
+// utility function, used by Split
```
maybe add a comment saying the context: `doing here because not to break compatibility; newer stake program ABI can just use newly-required Rent sysvar` or the like.

CriesofCarrots replied:
Added

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -969,6 +1007,17 @@ pub fn calculate_points(
+    source_data_len: u64,
+    split_data_len: u64,
+) -> u64 {
+    let lamports_per_byte_year =
+        source_meta.rent_exempt_reserve / (source_data_len + ACCOUNT_STORAGE_OVERHEAD);
+    lamports_per_byte_year * (split_data_len + ACCOUNT_STORAGE_OVERHEAD)
```
Also, let's cross-ref this out-of-wild rent calculation was unavoidable at `Rent` in rent.rs. :)

CriesofCarrots replied:
Where would you like to see that comment in rent.rs?

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -969,6 +1011,19 @@ pub fn calculate_points(
+//This emulates current Rent math in order to preserve backward compatibility. In the future, and
+//to support variable rent, the Split instruction should pass in the Rent sysvar instead.
+fn calculate_split_rent_exempt_reserve(
+    source_meta: &Meta,
```
nits: `source_rent_exempt_reserve`?

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -757,33 +761,71 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+                        // requested, less any lamports needed to cover the rent-exempt reserve
+                        (
+                            lamports,
+                            lamports - meta.rent_exempt_reserve.saturating_sub(split.lamports()?),
```
I think this must be `lamports - split_rent_exempt_reserve.saturating_sub(split.lamports()?)`, right?

CriesofCarrots replied:
Ah, yes, correct.

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -757,33 +761,71 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+                        // split account.
+                        (
+                            lamports - meta.rent_exempt_reserve,
+                            lamports - meta.rent_exempt_reserve,
```
Also, this must be `lamports - split_rent_exempt_reserve.rent_exempt_reserve`, right?

CriesofCarrots replied:
No, definitely not. The amount of stake delegated should not be able to increase by splitting to a smaller-sized account.

ryoqun replied:
Hhm. That's because of bad side-effect to the stake distribution for the current epoch? Maybe I'm lacking some precondition.

CriesofCarrots replied:
Hmm, it looks like the max stake delta would be ~69 SOL. Is that small enough we don't care?

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -757,33 +761,71 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+                    );
 
                     // verify enough lamports for rent in new stake with the split
                     if split.lamports()? + lamports < meta.rent_exempt_reserve
```
instead of `split_rent_exempt_reserve > meta.rent_exempt_reserve` below, the condition can be tweaked to `split.lamports()? + lamports < split_rent_exempt_reserve`?

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -757,33 +761,71 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+                    // verify that the same rent_exempt_reserve is sufficient to make the new
+                    // stake account rent exempt needed because the new uninitialized may be any
+                    // size larger than std::mem::size_of::<StakeState>()
+                        || split_rent_exempt_reserve > meta.rent_exempt_reserve
```
this new error condition will make it fail to split smaller-sized stake account into bigger-sized stake account even if there is no inconsistency between `rent_exempt_reserve` and `data_len()`. That will be very hard to debug for users. If could, I'd like to avoid this by adjusting split_stake and split_rent_exempt_reserve...

CriesofCarrots replied:
Yes, my intent was to prevent all splits from smaller-sized stake accounts into bigger-sized ones, since we are otherwise addressing the causes of small stake accounts with incorrect larger `rent_exempt_reserve`s.

> If could, I'd like to avoid this by adjusting split_stake and split_rent_exempt_reserve...

I'm sorry, I don't understand what you mean. Unless you are referring to your comment above?

CriesofCarrots commented:
@t-nelson @ryoqun , once more if you would 🙏 
I reworked this such that:
- The Split account `Meta::rent_exempt_reserve` is calculated and set to the correct value for the new account, regardless of whether it is higher or lower than the original
- Splits that would leave the new account not rent exempt fail
- Split account lamports = stake + rent_exempt_reserve, *except* if the account is prefunded with more than the rent_exempt_reserve, to prevent the magic activation of stake via split (which could be up to u64::MAX)
- Otherwise, the max possible change in activated stake by splitting between 2 accounts of dissimilar sizes is ~69 SOL

@ryoqun , hoping for a little more guidance about where you'd like documentation in rent.rs 🙇‍♀️ . Otherwise, I think I've addressed all of your comments.

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -757,31 +761,61 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+                    let (remaining_stake_delta, split_stake_amount) = if lamports
+                        == self.lamports()?
+                    {
+                        // If split amount equals the full source stake, the new split stake must
+                        // equal the same amount, regardless of any current lamport balance in the
+                        // split account. Since split accounts retain the state of their source
+                        // account, this prevents any magic activation of stake by prefunding the
+                        // split account.
```
:+1:

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -757,31 +761,61 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+                        // split account.
+                        (
+                            lamports - meta.rent_exempt_reserve,
+                            lamports - split_rent_exempt_reserve,
```
Argurably hard to notice, but I think this causes overflow, allowing user to stake around `u64::max_size()` if all of following conditions are met by specifically creating crafted account state and transaction by malicious users:

1. `lamports` is identical as `source.lamports()`.
2. `split` account is bigger than source; so `split_rent_exempt_reserve` > `meta.rent_exempt_reserve`
3. `split` account is pre-funded so that `split.lamports()? + lamports >= split_rent_exempt_reserve`.

ryoqun replied:
I think the best thing we could here is to reject such a case? Also, I'd like to add a test for this. (I know you already adding very extensive set of tests. Sorry for more trouble).

CriesofCarrots replied:
I think I see. I'll write a test to confirm. Yes, I think rejecting makes sense.

CriesofCarrots replied:
Please see `test_split_100_percent_of_source_to_larger_account_edge_case()`

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -757,31 +761,61 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
                     // verify enough lamports for rent in new stake with the split
-                    if split.lamports()? + lamports < meta.rent_exempt_reserve
-                     // verify enough lamports left in previous stake and not full withdrawal
+                    if split.lamports()? + lamports < split_rent_exempt_reserve
+                    // verify enough lamports left in previous stake and not full withdrawal
                         || (lamports + meta.rent_exempt_reserve > self.lamports()? && lamports != self.lamports()?)
```
nits: I think these conditions are bit hard to read? How about swapping this inequality around `lamports` like this?

```rust
if lamports < split_rent_excempt_reserve + lamports
    || (lamports > self.lamports()? - meta.rent_excempt_reserver && ...)
```

CriesofCarrots replied:
✅  Reframed all those checks in terms of `lamports` (and also let cargo fmt work on those lines)

ryoqun commented on programs/stake/src/stake_state.rs:
```diff
@@ -757,31 +761,61 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+                        // requested, less any lamports needed to cover the rent-exempt reserve
+                        (
+                            lamports,
+                            lamports - split_rent_exempt_reserve.saturating_sub(split.lamports()?),
```
I think similar attack is possible here like this: https://github.com/solana-labs/solana/pull/13357/files#r517344371

CriesofCarrots replied:
This case is covered by the initial lamports check (more obvious with your rewriting: `lamports < split_rent_excempt_reserve.saturating_sub(split.lamports()?)`)

CriesofCarrots replied:
I included my test that demonstrates this: `test_split_to_larger_account_edge_case()`

ryoqun commented:
Still pondering on this. But I think this getting very solid with many new tests. Thanks for great work!

ryoqun commented:
> @ryoqun , hoping for a little more guidance about where you'd like documentation in rent.rs bowing_woman . Otherwise, I think I've addressed all of your comments.

maybe, this fn:

https://github.com/solana-labs/solana/blob/b0d1ae1d8bac917198b8130be0464a1166f1b1fe/sdk/program/src/rent.rs#L49

along like:

```
// note that stripped down version of this calcuration is reluctantly copied into calculate_split_rent_exempt_reserve in the stake program
```

t-nelson commented on programs/stake/src/stake_state.rs:
```diff
@@ -3689,4 +4309,30 @@ mod tests {
+
+    #[test]
+    fn test_calculate_lamports_per_byte_year() {
+        let rent_exempt_reserve = 2282880;
```
Probably in the above tests as well

CriesofCarrots replied:
Probably a good idea, thanks!

CriesofCarrots replied:
Done

t-nelson reviewed (commented):
Looking good!  I just have a couple test hardening recommendations

CriesofCarrots commented:
This will need rebasing on #13394 (make sure it goes in stake_state.rs, as opposed to the legacy mod)

CriesofCarrots commented on programs/stake/src/stake_state.rs:
```diff
@@ -3689,4 +4309,30 @@ mod tests {
+            calculate_split_rent_exempt_reserve(rent_exempt_reserve, data_len, larger_data),
+            expected_rent_exempt_reserve
+        );
+
```
Added

t-nelson reviewed (dismissed):
LGTM!
